compaction/leveled_compaction_strategy.cc

```diff
@@ -69,7 +69,11 @@ compaction_descriptor leveled_compaction_strategy::get_major_compaction_job(tabl
 }
 
 void leveled_compaction_strategy::notify_completion(const std::vector<shared_sstable>& removed, const std::vector<shared_sstable>& added) {
-    if (removed.empty() || added.empty()) {
+    // All the update here is only relevant for regular compaction's round-robin picking policy, and if
+    // last_compacted_keys wasn't generated by regular, it means regular is disabled since last restart,
+    // therefore we can skip the updates here until regular runs for the first time. Once it runs,
+    // it will be able to generate last_compacted_keys correctly by looking at metadata of files.
+    if (removed.empty() || added.empty() || !_last_compacted_keys) {
```

Comment:
Maybe it would be better to engage _last_compacted_keys as soon as we switch to LCS. Or even make it non-optional.

Reply:
Perhaps better than improving _last_compacted_keys initialization, it's to rewrite LCS picking policy to promote file with least overlapping into next level. That can reduce WA.

I am still asking myself if I should rewrite LCS, or incrementally fix it. LCS needs to switch to run (deprecate sstable level metadata), increase fragment size (0.16M is suboptimal), switch to dynamic level to fix transient SA of 2, change the promotion policy. Given all changes needed, and today's implementation being a bag of heuristics, perhaps a rewrite is easier.

ICS + SAG is not in the same region of the efficiency space as classical leveled, it's a hybrid layout, somewhere in the middle of tiered and leveled.

@avikivity what do you think?

Reply:
Round robin promotion policy is clearly suboptimal. By switching to least overlapping policy, we can get rid of this last_compacted_keys mess.